src/Microsoft.Azure.SignalR/EndpointRouters/DefaultEndpointRouter.cs

```diff
@@ -45,5 +46,49 @@ private ServiceEndpoint[] GetNegotiateEndpoints(IEnumerable<ServiceEndpoint> end
 
             return secondary;
         }
+
+        /// <summary>
+        ///  Choose endpoint randomly by weight. 
+        ///  The weight is defined as the remaining connection quota.
+        ///  The least weight is set to 1. So instance with no connection quota still has chance.
+        /// </summary>
+        private ServiceEndpoint GetEndpointAccordingToWeight(ServiceEndpoint[] availableEndpoints)
+        {
+            //first check if weight is available or necessary
+            if (availableEndpoints.Any(endpoint => endpoint.EndpointMetrics.ConnectionCapacity == 0) ||
+                availableEndpoints.Length == 1)
+            {
+                return GetEndpointRandomly(availableEndpoints);
+            }
+
+            var we = new int[availableEndpoints.Length];
+            var totalCapacity = 0;
+            for (var i = 0; i < availableEndpoints.Length; i++)
+            {
+                var endpointMetrics = availableEndpoints[i].EndpointMetrics;
+                var remain = endpointMetrics.ConnectionCapacity -
+                             (endpointMetrics.ClientConnectionCount +
+                              endpointMetrics.ServerConnectionCount);
+                var weight = remain > 0 ? remain : 1;
+                totalCapacity += weight;
+                we[i] = totalCapacity;
+            }
+
+            var index = StaticRandom.Next(totalCapacity);
+            for (var i = 0; i < we.Length; i++)
+            {
+                if (we[i] > index)
+                {
+                    return availableEndpoints[i];
+                }
+            }
```

Comment:
`return available[Array.FindLastIndex(we, x => x < index) + 1]`?

Reply:
simplified